DynRPG whitelist + easyrpg_raw command

jetrotal opened the pull request:
Whitelist selected dynRPG commands, so people can use them without enabling the dynRPG plugin.
![image](https://github.com/EasyRPG/Player/assets/45118493/9feab415-2ab5-4e7f-9336-3100ba5ce8b0)

------------------------------

Invoke a command using raw input, similar to how maniacs patch raw work.
easyRPG raw syntax:
```js
@easyrpg_raw 10110, "test", [1,2,3,4,V1],1 
//Message box
@easyrpg_raw $ShowMessage, T1, [1,2,3,4,V1],1
//Also Message box
```

jetrotal committed:
whitelist DynRPG Commands

whitelisting easyRPG commands so they can be called even when dynRPG is off.

jetrotal committed:
@easyrpg_raw DynRPG command

Invoke a command using raw values, similar to how maniacs patch raw work.
NOTE FIXME PLS.

Ghabry committed:
DynRpg: Refactor to use a Game_DynRpg instance.

This makes cleanup of state easier when switching between games.

Ghabry committed:
DynRpg: easyrpg_raw rewrite and fix memory issues

Ghabry commented:
We need to parse deeper

![grafik](https://github.com/EasyRPG/Player/assets/1331889/55b71947-cfbe-466c-bf26-ee4e3471c5bb)

Ghabry committed:
DynRPG: Support T (Maniac Patch string variable) in the token parser

Ghabry commented:
The crossover nobody asked for

![grafik](https://github.com/EasyRPG/Player/assets/1331889/4fe9863a-22b1-4830-a29b-2d23d7782805)

Ghabry commented:
btw that "frame pushing" looks like a safe solution for this ``Call Command`` thing of Maniacs. Is this event command hard to implement?

jetrotal commented:
> btw that "frame pushing" looks like a safe solution for this `Call Command` thing of Maniacs. Is this event command hard to implement?

https://jetrotal.github.io/CSA/#MNC%20|%20Call%20Command
![image](https://github.com/EasyRPG/Player/assets/45118493/f9d7d333-ba8c-4135-bfe5-c7ebc9bcb88d)


It's similar but uses an array of variables in sequence for the arguments

Ghabry commented:
That PR is kinda the wrong location for this, so one final question before we move this maybe somewhere else:

I think that CSA example needs more permutations and a 3rd argument to figure out how this works. E.g. 

``@cmd 1, "", .args v[2], 3`` gives parameters: ``0, 1, 2, 3``. How is ``v[2]`` encoded? o_O

jetrotal commented:
@Ghabry I also set up a collection of useful constants, that came from destinyScript. Those could be useful for parsing common values:
![image](https://github.com/EasyRPG/Player/assets/45118493/34ff259c-0e73-4254-a2cf-c98df1591191)


made in two formats, the second one is a Map, I guess that's the one that would work with current code:
https://www.diffchecker.com/YAS7V0Mi/

copy the text at the bottom.

jetrotal committed:
DynRPG  @Raw - Calling Events by name.

@easyrpg_raw @ShowMessage, "This is new"

jetrotal commented:
![image](https://github.com/EasyRPG/Player/assets/45118493/d9853bd0-38f7-46f4-8362-fc763a0c28d9)
![image](https://github.com/EasyRPG/Player/assets/45118493/90481943-ee98-4445-8542-42be0bca1c5d)

jetrotal committed:
DynRPG Raw - fill numeric parameters.

jetrotal committed:
DynRPG - easyrpg_raw: Numeric Parameters can call destiny constants

implementation is a bit ugly, need help to clear this

jetrotal commented:
![image](https://github.com/EasyRPG/Player/assets/45118493/27cab190-96f5-4bb7-8884-9f633256297f)

jetrotal committed:
DynRPG Easyrpg_raw - simplify commands

removed repeated code, and overcomplex logic.

jetrotal commented:
Hey guys, Removed a lot of repeated code, including using some functions we already have in Utils.

I think this is good enough for a working raw command interpreter.
I wanted to break it into multiple lines of code, but Ghabry suggested writing in the future an eval command for that
(maybe `easyrpg_eval "prefix", "string"` ?)

Other things that may be important for this in the future is setting up the "constants.h" content as
auto-generated entries from liblcf.

--------------------------------------

As soon as this this pr is aproved, I can rework #3124 as you guys asked.

jetrotal committed:
Multiline interpreter + major refactor

Managed to make it work, but the code is a bit messy and its syntax is not very flexible;

jetrotal commented:
Ok, multi-line is done. 
I had to add something similar to the "," on the part of the code that parses text to arg[n]:
It parses ';' as ',' too with an extra input in case of being "easyrpg_raw". I hope that's not too invasive...

Ghabry commented:
I talked with Jetrotal: I will help getting this in when I'm done with my remaining issues for 0.8.1 milestone. Likely in one or two weeks.

My current plan is to split it into two PRs. One for the DynRPG refactor, one for easyrpg_raw. Makes reviewing this a bit easier :)

jetrotal committed:
easyrpg_raw: rebase and adapt old code + preparations for future features

Ghabry commented:
Would you mind when I swap the argument order to:

```
$ControlVars "A", 1, 0, 3, 4, 6, 8"
```

Makes "indent" the 2nd parameter and the numeric list everything from 3rd to the end. Makes the parsing simpler and is not ambigious.

jetrotal commented:
> Would you mind when I swap the argument order to:
> 
> ```
> $ControlVars "A", 1, 0, 3, 4, 6, 8
> ```
> 
> Makes "indent" the 2nd parameter and the numeric list everything from 3rd to the end. Makes the parsing simpler and is not ambigious.

hm... I made it this way to let people ignore the indent parameter, and read each type of parameter by group

jetrotal commented:
@Ghabry asked for me to post the event2string command I was working on, here it is:
https://gist.github.com/jetrotal/e8b35118b94ddbe8ee10885e220f706c

jetrotal committed:
New Command: 2099 - StoreCommands

Store commands inside a Stringvariable.

//$storeCommands "preffix",[evtType_isVar, evtType, evtId_isVar, evtId, evtPage_isVar, evtPage, targetStrVar_isVar, targetStrVar]

// 2099, "@easyrpg_raw", [1,4, 1,5, 1,6, 1,7], 0;

jetrotal committed:
StoreCommands (2099) - Use kCodeTags.tag instead of custom tag reader

jetrotal committed:
@easyrpg_raw - Further use kCodeTags instead of a custom tag reader

now, our raw code also use this library.
